detect_cans_in_fridge_201202/euslisp/fridge-demo-smach-main.l

```diff
@@ -1,198 +1,129 @@
+    (if res
+      (set-alist 'description (format nil "~Aを持っていったよ" can-type) userdata)
+      (set-alist 'description (format nil "~Aを持っていけなかったよ" can-type) userdata))
+    (if *image* (set-alist 'image (remove #\newline (base64encode (send *image* :serialize))) userdata))
+    res))
+
+
+(defun fridge-demo-state-machine ()
   (setq *sm*
         (make-state-machine
-         '((:goto-front-of-fridge -> :open-fridge-door)
-           (:open-fridge-door ->  :grasp-can)
-           (:grasp-can -> :close-fridge)
-           (:close-fridge -> :success))
-         '((:goto-front-of-fridge *goto-fridge-sm*)
-           (:open-fridge-door *open-fridge-sm*)
-           (:grasp-can *grasp-can-sm*)
-           (:close-fridge *close-fridge-sm*))
-         '(:goto-front-of-fridge)
-         '(:success))))
-
-(defun init ()
-  (ros::roseus "sample_parallel_state_machine")
-  (setq *use-arm* :larm)
-  (setq *fridge-demo* (instance fridge-demo-machine :init))
-  (make-goto-front-of-fridge)
-  (make-open-fridge-door-state-machine)
-  (make-grasp-can-state-machine)
-  (make-close-fridge-state-machine)
-  (make-fridge-demo-state-machine)
-  (ros::ros-info "created state machine ~A" *sm*))
-
-(warning-message 3 ";; (init)~%")
-
-(defun demo (&optional (wait t))
-  (setq *wait* wait)
-  (when (or (not (boundp '*sm*)) (not (boundp '*fridge-demo*)))
-    (init))
-  (exec-smach-with-spin *sm*))
-
-(warning-message 3 ";; (demo)~%")
-
-(provide :fridge-demo-machine)
+          '((:init -> :move-to-fridge)
+            (:move-to-fridge -> :open-fridge)
+            (:move-to-fridge !-> nil)
+            (:open-fridge -> :grasp-can)
+            (:open-fridge !-> nil)
+            (:grasp-can -> :close-fridge)
+            (:grasp-can !-> nil)
+            (:close-fridge -> :move-to-initial-position)
+            (:close-fridge !-> nil)
+            (:move-to-initial-position -> t)
+            (:move-to-initial-position !-> nil))
+          '((:init 'init-func)
+            (:move-to-fridge 'move-to-fridge-func)
+            (:open-fridge 'open-fridge-func)
+            (:grasp-can 'grasp-can-func)
+            (:close-fridge 'close-fridge-func)
+            (:move-to-initial-position 'move-to-initial-position-func))
+          '(:init)
+          '(t nil))))
+
+
+(defun fridge-demo-smach-main (&key (can-type "georgia")
+                                    (finish-type :pass-to-human)
+                                    (use-arm :larm)
+                                    (use-base nil))
+  (when (not (boundp '*sm*))
+    (fridge-demo-state-machine))
+  (let ((result-state
+          (exec-state-machine
+            *sm* `((description . ,(format nil "冷蔵庫に~Aを取りに行ったよ" can-type))
+                   (image . "")
+                   (can-type . ,can-type)
+                   (finish-type . ,finish-type)
+                   (use-arm . ,use-arm)
+                   (use-base . ,use-base))
+            :hz 2.0)))
```

Comment:
Perhaps pr2's smach_to_mail.py is up and running with this node name.
```suggestion
            :srv-name "smach_to_mail"
            :hz 2.0)))
```

Reply:
https://github.com/jsk-ros-pkg/jsk_robot/blob/6a352a709b5964a419f0a4a33059be49e5c00a10/jsk_pr2_robot/jsk_pr2_startup/pr2.launch#L243-L246

Reply:
but if demo's smach srv-name is set to smach_to_mail_server, it may be complicated.
It may be a good idea to change the namespace to receive dynamically in smach_to_mail.py.

Reply:
smach topic name should start from `server_name` because `smach_viewer`'s default value is `server_name`.
and i think i tried renaming `smach_to_mail` node to `server_name` last night, which was not successful.
i will check it again.